### ISSUE TITLE
Precompiled headers not working on linux/macOS in latest MOC branch

### ISSUE THREAD

**Issue by SRSaunders** [RobertBeckebans/RBDOOM-3-BFG#931]
PCH for linux/macOS fails in your new MOC branch.  This is caused by updating to cmake 3.10 policies.

A simple solution is to add the following lines to the main CMakeLists file:

```
    cmake_minimum_required(VERSION 3.10)

--> # SRS - Use old policy and treat DEFINITIONS as a built-in directory property
--> #     - This is needed to keep precompiled headers working on Linux and macOS
--> cmake_policy(SET CMP0059 OLD)

...
```

Also - there are macOS compile failures due to missing definitions for `__cpuidex` and `_xgetbv` on macOS, as well as trying to use `memalign()` vs. `posix_memalign()`.  I would be happy to add a PR to solve once you merge into master.

Otherwise the MOC code works well with a nice performance increase on macOS/x86.  I have not yet tested on Apple Silicon but I presume the MOC code will be disabled there with no change in perf.

**Comment by RobertBeckebans** [comment]
Strange I tested the branch with precompiled headers on on Fedora 40 and Kubuntu 23.10 with both clang and gcc and it worked for me.

**Comment by SRSaunders** [comment]
I wonder if there is a difference in cmake policy behaviour between platforms?  It certainly doesn't work on macOS when using cmake minimum 3.10 unless I activate the old policy for CMP0059 re the meaning of `DEFINITIONS`.  This policy changed in the cmake 3.3 version.  Linux/macOS precompiled headers depend on `DEFINITIONS` working in the cmake line: `GET_DIRECTORY_PROPERTY(_directory_flags DEFINITIONS)`.

Also, I noticed that your **cmake-linux-release.sh** script currently contains:

`cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DONATIVE=ON -DFFMPEG=OFF -DBINKDEC=ON -**DUSE_PRECOMPILED_HEADERS=OFF** -DUSE_INTRINSICS_SSE=ON ../neo`

So something must have been going on for you to set that PCH compile option.

**Comment by RobertBeckebans** [comment]
I tested it with ON/OFF and did some reformatting for the cmake precompiled headers code. I also ran into some issues with existing precompiled header files although I did =OFF as an option. I probably will put another rm -f command into the cmake-linux-*.sh files so it makes sure there are no precompiled leftovers in the neo/ folder.

I'm going to check this again next week.

**Comment by SRSaunders** [comment]
Sounds good.  I will test PCH on my Manjaro Linux setup to compare notes.

**Comment by SRSaunders** [comment]
Tested on Manjaro linux with gcc & clang using cmake 3.29.3:  PCH **doesn't work** in your current MOC branch and needs my patch above, at least in the short term.

However,  since this is a deprecated policy, once your work is merged I could look into changing the CMakeLists code so linux/macOS PCH works with current cmake default policies, i.e. remove use of `DEFINITIONS` and the policy override.

**Comment by SRSaunders** [comment]
Thanks for merging the solution.  Closing this issue.
